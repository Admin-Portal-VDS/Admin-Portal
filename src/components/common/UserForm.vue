<template>
  <div class="form-container">
    <form @submit.prevent="handleSubmit">
      <!-- Basic Info Section -->
      <div class="section">
        <h2>Basic Info</h2>
        <div
          v-for="field in basicInfoFields"
          :key="field.id"
          class="form-group"
        >
          <label :for="field.id">{{ field.label }}</label>
          <input
            :id="field.id"
            v-model="form[field.model]"
            :type="field.type"
            :placeholder="field.placeholder"
            :required="field.required"
            :data-id="field.id"
            class="full-width-input"
          />
        </div>

        <!-- Select Group(s) Dropdown -->
        <div class="form-group">
          <label for="groups">Select Group(s)</label>
          <select
            data-id="groups"
            id="groups"
            v-model="form.groups"
            class="select-group"
          >
            <option value="">Select Group(s)</option>
            <option
              v-for="group in groups"
              :key="group.value"
              :value="group.value"
              :data-id="group.value"
            >
              {{ group.label }}
            </option>
          </select>
        </div>
      </div>

      <div class="divider">
        <span></span>
      </div>

      <!-- Settings and Permissions Section -->
      <div class="section">
        <h2>Settings and Permissions</h2>
        <div class="user-type">
          <select
            data-id="userType"
            id="userType"
            v-model="form.userType"
            required
          >
            <option
              v-for="option in userTypeOptions"
              :key="option.value"
              :value="option.value"
              :data-id="option.value"
            >
              {{ option.text }}
            </option>
          </select>
        </div>
      </div>

      <!-- User Type Section (Moved after Settings and Permissions) -->
      <div class="section">
        <div
          v-for="(label, key) in permissionFields"
          :key="key"
          class="form-group"
        >
          <label>
            <input
              type="checkbox"
              v-model="form.permissions[key]"
              :data-id="`permission-${key}`"
            />
            {{ label }}
          </label>
        </div>
      </div>

      <div class="divider">
        <span></span>
      </div>

      <!-- Extensions Section -->
      <div class="section">
        <h2>Extensions</h2>
        <div
          v-for="(label, key) in extensionOptions"
          :key="key"
          class="form-group"
        >
          <label>
            <input
              type="radio"
              v-model="form.extensionOption"
              :value="key"
              :data-id="`extension-${key}`"
            />
            {{ label }}
          </label>
        </div>

        <div
          data-id="notice"
          v-if="form.extensionOption === 'purchase'"
          class="notice"
        >
          You will be brought to the store to purchase an extension after saving
          the user. Once you have purchased an extension, go to the Extensions
          page to assign the extension to this user.
        </div>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  data() {
    return {
      form: {
        firstName: '',
        lastName: '',
        username: '',
        email: '',
        confirmEmail: '',
        groups: '',
        userType: '',
        permissions: {
          serviceAnnouncements: false,
          webAccess: true,
          sendEmail: true,
          manageCallerId: false,
          manageCallBlocking: false,
        },
        extensionOption: '',
      },
      basicInfoFields: [
        {
          id: 'firstName',
          model: 'firstName',
          type: 'text',
          label: 'First Name*',
          placeholder: 'Enter first name',
          required: true,
        },
        {
          id: 'lastName',
          model: 'lastName',
          type: 'text',
          label: 'Last Name*',
          placeholder: 'Enter last name',
          required: true,
        },
        {
          id: 'username',
          model: 'username',
          type: 'text',
          label: 'Username*',
          placeholder: 'Enter username',
          required: true,
        },
        {
          id: 'email',
          model: 'email',
          type: 'email',
          label: 'Email*',
          placeholder: 'Enter email',
          required: true,
        },
        {
          id: 'confirmEmail',
          model: 'confirmEmail',
          type: 'email',
          label: 'Confirm Email*',
          placeholder: 'Confirm email',
          required: true,
        },
      ],
      groups: [
        { value: 'group1', label: 'Group 1' },
        { value: 'group2', label: 'Group 2' },
        { value: 'group3', label: 'Group 3' },
      ],
      userTypeOptions: [
        { value: '', text: 'Select User Type' },
        { value: 'accountAdmin', text: 'V Account Administrator' },
        { value: 'billingAdmin', text: 'Billing Admin' },
        { value: 'callQueuesAdmin', text: 'Call Queues Admin' },
        {
          value: 'callQueuesReportingAdmin',
          text: 'Call Queues Reporting Admin',
        },
        { value: 'endUser', text: 'End User' },
        { value: 'endUserNoDashboard', text: 'End User No Dashboard' },
        { value: 'reportsAdmin', text: 'Reports Admin' },
        { value: 'usersAdmin', text: 'Users Admin' },
      ],
      permissionFields: {
        serviceAnnouncements:
          'Update this user with important service announcements',
        webAccess: 'Allow web and mobile app access',
        sendEmail: 'Send new user email',
        manageCallerId:
          'Allow this user to set and block the Outbound Caller ID for their extension(s)',
        manageCallBlocking:
          'Allow this user to manage Call Blocking for their extension(s)',
      },
      extensionOptions: {
        assign: 'Assign an available extension',
        purchase: 'Purchase a new extension',
        none: 'User does not need an extension',
      },
    };
  },
  methods: {
    handleSubmit() {
      console.log('Form submitted:', this.form);
    },
  },
};
</script>

<style scoped>
/* Reset margin and padding for the body */
body {
  background-color: #f5f5f5;
  font-family: Arial, sans-serif;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.form-container {
  max-width: 800px;
  width: 100%; /* Make it responsive */
  padding: 20px;
  border-radius: 8px;
  margin-left: 400px;
}
.select-group {
  height: 45px;
  width: 180px;
  margin-top: 10px;
}
.section {
  margin-bottom: 20px;
}
.full-width-input {
  width: 500px;
  height: 20px;
  color: black;
  border: 1px solid rgba(9, 8, 8, 0.64);
  margin-bottom: 10px;
  margin-top: 10px;
}

h2 {
  font-size: 1.5rem;
  color: #333;
  margin-top: 25px;
  margin-bottom: 15px;
}

.form-group {
  margin-bottom: 15px;
}
.user-type {
  margin-top: 25px;
  width: 80px;
}

label {
  display: block;
  font-weight: bold;
  margin-bottom: 5px;
}

input,
select {
  padding: 10px;
  font-size: 1rem;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.notice {
  font-size: 0.9rem;
  color: #d9534f;
  background-color: #f9e2e2;
  padding: 10px;
  border: 1px solid #d6b3b3;
  border-radius: 4px;
}
.divider {
  display: flex;
  align-items: center;
  margin: 20px 0;
  color: #888;
}
.divider::before,
.divider::after {
  content: '';
  flex: 1;
  height: 1px;
  background: #ddd;
  margin: 0 0px;
}
/* Responsive Design */
@media (max-width: 768px) {
  .form-container {
    padding: 15px;
  }
}

@media (max-width: 480px) {
  .form-container {
    padding: 10px;
  }
}
</style>
