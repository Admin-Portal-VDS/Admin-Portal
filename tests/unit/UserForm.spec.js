import { mount } from '@vue/test-utils';
import UserForm from '@/components/common/UserForm';

describe('UserForm.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(UserForm, {});
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders the form with all sections and inputs', () => {
    expect(wrapper.find('form').exists()).toBe(true);

    const basicInfoFields = wrapper.vm.basicInfoFields;
    basicInfoFields.forEach((field) => {
      expect(wrapper.find(`label[for="${field.id}"]`).exists()).toBe(true);
      expect(wrapper.find(`#${field.id}`).exists()).toBe(true);
    });
    expect(wrapper.find('[data-id="groups"]').exists()).toBe(true);
    expect(wrapper.find('[data-id="userType"]').exists()).toBe(true);
  });

  test('renders permissions checkboxes', () => {
    const permissionFields = wrapper.vm.permissionFields;

    Object.keys(permissionFields).forEach((key) => {
      const label = wrapper
        .findAll('label')
        .filter((node) => node.text().includes(permissionFields[key]))[0];

      expect(label.exists()).toBe(true);

      const checkbox = label.find(`[data-id="permission-${key}"]`);
      expect(checkbox.exists()).toBe(true);
    });
  });

  test('renders extension options correctly', () => {
    const extensionOptions = wrapper.vm.extensionOptions;

    Object.keys(extensionOptions).forEach((key) => {
      expect(wrapper.find(`[data-id="extension-${key}"]`).exists()).toBe(true);
    });
  });

  test('displays notice when extension option is "purchase"', async () => {
    await wrapper.setData({
      form: { ...wrapper.vm.form, extensionOption: 'purchase' },
    });

    const notice = wrapper.find('[data-id="notice"]');
    expect(notice.exists()).toBe(true);
    expect(notice.text()).toContain(
      'You will be brought to the store to purchase an extension',
    );
  });

  test('validates required fields', async () => {
    const formFields = [
      'firstName',
      'lastName',
      'username',
      'email',
      'confirmEmail',
    ];

    formFields.forEach((field) => {
      const input = wrapper.find(`#${field}`);
      expect(input.attributes('required')).toBeDefined();
    });

    const userType = wrapper.find('[data-id="userType"]');
    expect(userType.attributes('required')).toBeDefined();
  });

  test('handles form submission', async () => {
    const handleSubmitMock = jest.fn();
    wrapper.vm.handleSubmit = handleSubmitMock;

    await wrapper.find('form').trigger('submit.prevent');
    expect(handleSubmitMock).toHaveBeenCalled();
  });

  test('updates v-model when input changes', async () => {
    const input = wrapper.find('[data-id="firstName"]');
    await input.setValue('John');
    expect(wrapper.vm.form.firstName).toBe('John');
  });

  test('updates selected group in form data', async () => {
    const select = wrapper.find('[data-id="groups"]');
    await select.setValue('group2');
    expect(wrapper.vm.form.groups).toBe('group2');
  });

  test('updates user type in form data', async () => {
    const select = wrapper.find('[data-id="userType"]');
    await select.setValue('billingAdmin');
    expect(wrapper.vm.form.userType).toBe('billingAdmin');
  });

  test('matches snapshot', () => {
    const wrapper = mount(UserForm);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
