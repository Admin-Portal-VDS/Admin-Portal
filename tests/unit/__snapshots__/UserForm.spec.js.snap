// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormComponent.vue matches snapshot 1`] = `
<div class="form-container">
  <form>
    <!-- Basic Info Section -->
    <div class="section">
      <h2>Basic Info</h2>
      <div class="form-group"><label for="firstName">First Name*</label><input id="firstName" type="text" placeholder="Enter first name" required="" class="full-width-input"></div>
      <div class="form-group"><label for="lastName">Last Name*</label><input id="lastName" type="text" placeholder="Enter last name" required="" class="full-width-input"></div>
      <div class="form-group"><label for="username">Username*</label><input id="username" type="text" placeholder="Enter username" required="" class="full-width-input"></div>
      <div class="form-group"><label for="email">Email*</label><input id="email" type="email" placeholder="Enter email" required="" class="full-width-input"></div>
      <div class="form-group"><label for="confirmEmail">Confirm Email*</label><input id="confirmEmail" type="email" placeholder="Confirm email" required="" class="full-width-input"></div><!-- Select Group(s) Dropdown -->
      <div class="form-group"><label for="groups">Select Group(s)</label><select id="groups" class="full-width-input">
          <option value="">Select Group(s)</option>
          <option value="group1">Group 1</option>
          <option value="group2">Group 2</option>
          <option value="group3">Group 3</option>
        </select></div>
    </div>
    <div class="divider"><span></span></div><!-- Settings and Permissions Section -->
    <div class="section">
      <h2>Settings and Permissions</h2>
      <div class="form-group"><label for="userType">User Type*</label><select id="userType" required="">
          <option value="">Select User Type</option>
          <option value="accountAdmin">V Account Administrator</option>
          <option value="billingAdmin">Billing Admin</option>
          <option value="callQueuesAdmin">Call Queues Admin</option>
          <option value="callQueuesReportingAdmin">Call Queues Reporting Admin</option>
          <option value="endUser">End User</option>
          <option value="endUserNoDashboard">End User No Dashboard</option>
          <option value="reportsAdmin">Reports Admin</option>
          <option value="usersAdmin">Users Admin</option>
        </select></div>
    </div><!-- User Type Section (Moved after Settings and Permissions) -->
    <div class="section">
      <div class="form-group"><label><input type="checkbox"> Update this user with important service announcements</label></div>
      <div class="form-group"><label><input type="checkbox"> Allow web and mobile app access</label></div>
      <div class="form-group"><label><input type="checkbox"> Send new user email</label></div>
      <div class="form-group"><label><input type="checkbox"> Allow this user to set and block the Outbound Caller ID for their extension(s)</label></div>
      <div class="form-group"><label><input type="checkbox"> Allow this user to manage Call Blocking for their extension(s)</label></div>
    </div>
    <div class="divider"><span></span></div><!-- Extensions Section -->
    <div class="section">
      <h2>Extensions</h2>
      <div class="form-group"><label><input type="radio" value="assign"> Assign an available extension</label></div>
      <div class="form-group"><label><input type="radio" value="purchase"> Purchase a new extension</label></div>
      <div class="form-group"><label><input type="radio" value="none"> User does not need an extension</label></div>
      <!--v-if-->
    </div>
  </form>
</div>
`;
